### PR TITLE
add poetry dataset setup

### PR DESCRIPTION
**Pull request by CheckMC**
Dataset Description
This dataset contains around 14,000 poems from the PoetryFoundation.org site. They are converted to question:response pairs, using the tags as topics.
5% of the dataset is titling requests -- the user provides a poem and asks the assistant to title it.

Languages
English

Dataset Structure
This dataset follows the OA format, which is:

INSTRUCTION (string): The user asks for a poem (from a variety of premade prompts) with topics (tags). If the given poem has no tags, the user asks for a poem on it's own.

RESPONSE (string): The assistant replies with the poem and title (from a variety of premade prompts).

SOURCE (string): The source is PoetryFoundation.org and the poet's name.

METADATA (JSON String):
{"author": "author of the original poem",
"title": "title of the poem",
"tags": "tags from poetry foundation."}

Preparing the Dataset
The dataset can be created with prepare.py. Make sure to install the required libraries in requirements.txt!

Contributions
Created by Check
Original dataset source - https://www.kaggle.com/datasets/tgdivy/poetry-foundation-poems

You can view it on my huggingface here: https://huggingface.co/datasets/checkai/instruction-poems
(this time i ran pre-commit so it should be good :D )

### CODE REVIEW

File: data/datasets/poetry_instruction/prepare.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+import pandas as pd
+
+# Read the CSV file into a pandas DataFrame. The CSV file was downloaded from Kaggle, and the source is linked in the README.
+
+CURRENT_DIR = os.path.dirname(__file__)  # Gets directory path of the current python module
+df = pd.read_csv(os.path.join(CURRENT_DIR, "PoetryFoundationData.csv"))
+
+# The data in the CSV file is not formatted correctly, so we need to clean it up.
+df["Title"] = df["Title"].replace("\n", "", regex=True).replace("\r", "", regex=True)
+df["Title"] = df["Title"].str.strip()
```
Comment (sedthh) [inline data/datasets/poetry_instruction/prepare.py:14]: you can also do something like
`df["Title"].apply(lambda x: "".join(x.splitlines()).strip())`

File: data/datasets/poetry_instruction/prepare.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+df["Poem"] = df["Poem"].str.strip()
+df["Poem"] = df["Poem"].str.replace("Translated from the French", "")
+
+# "writing_prompts" are for tasks requesting the assistant to write a poem.
+# "topic" or "notTopic" are used depending if the original dataset had a topic listed for the poem or not.
+writing_prompts_topic = [
+    "Write me a poem about $topic.",
+    "I want a poem about $topic.",
+    "Can you write a poem? Make it about $topic.",
+    "Compose a poem, about $topic.",
```
Comment (sedthh) [inline data/datasets/poetry_instruction/prepare.py:25]: Multiple prompts, well done!

**Review by olliestanley — COMMENTED** [review]
Will need to remove the output CSV from the Git repo

File: data/datasets/poetry_instruction/prepare.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+# Loop through the writing tasks and process them.
+for index, row in writing_tasks.iterrows():
+    # Get data from the entry
+    poem = row["Poem"]
+    topics = row["Tags"]
+    title = row["Title"]
+    author = row["Poet"]
+
+    # Variables to store to instruction, reply, source, and metadata.
+    instruction = random.choice(writing_prompts_topic).replace("$topic", str(topics))
```
Comment (sedthh) [inline data/datasets/poetry_instruction/prepare.py:112]: you can also do string.format()  with "blabla {topic}" like https://www.w3schools.com/python/ref_string_format.asp

but this is okay too

File: data/datasets/poetry_instruction/prepare.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    reply = random.choice(replies_topic).replace("$topic", str(topics)).replace("$title", title).replace("$poem", poem)
+    source = "PoetryFoundation.org" + " - " + author
+    metadata = {"author": author, "title": title, "tags": str(topics), "task_type": "writing"}
+
+    # If the entry has "nan" as a topic, change the prompts to the non-topic ones.
+    if " nan" in instruction:
```
Comment (sedthh) [inline data/datasets/poetry_instruction/prepare.py:118]: please use `pd.isna(instruction)` or something similar, the "nan" is an artifact from converting NaN (missing/None) to string

File: data/datasets/poetry_instruction/requirements.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+pandas==2.0.0
```
Comment (sedthh) [inline data/datasets/poetry_instruction/requirements.txt:1]: Well done

File: data/datasets/poetry_instruction/prepare.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    # If the entry has a topic, use the topic to generate a writing prompt and a reply.
+
+    entry = {"INSTRUCTION": instruction, "RESPONSE": reply, "SOURCE": source, "METADATA": json.dumps(metadata)}
+    prepared_data.append(entry)
+
+# Convert prepared_data to a DataFrame.
+prepared_data = pd.DataFrame(prepared_data)
+
+# Save the DataFrame to disk in the Parquet format
+prepared_data.to_parquet("output.parquet", row_group_size=100, engine="pyarrow", index=False)
```
Comment (sedthh) [inline data/datasets/poetry_instruction/prepare.py:148]: This is also correct.

**Review by sedthh — CHANGES_REQUESTED** [review]
Added some minor comments, but otherwise this is very well done. You have added all the fixes requested previously. 

Please add the dataset to the parent `__init__.py` and it should be good to go!

File: data/datasets/poetry_instruction/prepare.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
+import json
+import os
+import random
+
+import pandas as pd
+
+# Read the CSV file into a pandas DataFrame. The CSV file was downloaded from Kaggle, and the source is linked in the README.
+
+CURRENT_DIR = os.path.dirname(__file__)  # Gets directory path of the current python module
```
Comment (sedthh) [inline data/datasets/poetry_instruction/prepare.py:9]: mabye you can also add a KAGGLE dataset loader to unpack this (I can give you an example code for this)

**Review by sedthh — CHANGES_REQUESTED** [review]
some more comments

**Review by olliestanley — APPROVED** [review]

**Pull request merged**